internal/kaas/cluster.go

```diff
@@ -0,0 +1,166 @@
+		} else {
+			if clientErr.ErrorMessage == clientError.ResourceNotFound {
+				return nil, clientError.NewClientError(clientErr, clientError.ResourceNotFound, fmt.Sprintf("Could not find cluster %s", name))
+			} else {
+				return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, fmt.Sprintf("Error getting cluster %s", name))
+			}
+		}
+	}
+
+	err = ValidateClusterComponents(clusterAPICR)
+	if err != nil {
+		return nil, clientError.NewClientError(err, clientError.InvalidConfiguration, fmt.Sprintf("Cluster %s have an invalid configuration", name))
+	}
+
+	cluster := &Cluster{}
+	err = cluster.GetClusterProperties(clusterAPICR)
+	if err != nil {
+		clientErr, ok := err.(*clientError.ClientError)
+		if !ok {
+			return nil, clientError.NewClientError(err, clientError.UnexpectedError, fmt.Sprintf("An Unexpected error happened while reading cluster %s properties", name))
+		} else {
+			if clientErr.ErrorMessage == clientError.InvalidConfiguration {
+				return nil, clientError.NewClientError(clientErr, clientError.InvalidConfiguration, fmt.Sprintf("Cluster %s is invalid due to missing or invalid labels", name))
+			}
+		}
+	}
+
+	return cluster, nil
+}
+
+func ListClusters(k *k8s.Kubernetes) ([]*Cluster, error) {
+
+	var clusterList []*Cluster
+
+	clusterListAPICR, err := k.ListClusters()
+	if err != nil {
+		clientErr, ok := err.(*clientError.ClientError)
+		if !ok {
+			return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, "Error listing clusters")
+		} else {
+			if clientErr.ErrorMessage == clientError.ResourceNotFound {
+				return nil, clientErr
+			} else if clientErr.ErrorMessage == clientError.EmptyResponse {
+				return nil, clientErr
+			} else {
+				return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, "Something went wrong when listing clusters")
+			}
+		}
+	}
+
+	for _, clusterAPICR := range clusterListAPICR.Items {
+		cluster := &Cluster{}
+		err = ValidateClusterComponents(&clusterAPICR)
+		if err != nil {
+			log.Printf("Skipping cluster %s because of invalid configuration: %s", cluster.Name, err.Error())
+			continue
+		}
+		err = cluster.GetClusterProperties(&clusterAPICR)
+		if err != nil {
+			clientErr, ok := err.(*clientError.ClientError)
+			if !ok {
+				log.Printf("An Unexpected error happened while reading cluster %s properties: %s", clusterAPICR.Name, err.Error())
+			} else {
+				if clientErr.ErrorMessage == clientError.InvalidConfiguration {
+					log.Printf("Cluster %s is invalid due to missing or invalid labels: %s", clusterAPICR.Name, err.Error())
+				}
+			}
+			continue
+		}
+		clusterList = append(clusterList, cluster)
+	}
+
+	if len(clusterList) == 0 {
+		return nil, clientError.NewClientError(nil, clientError.EmptyResponse, "no valid clusters were found, some clusters have invalid configuration")
```

Comment:
Should start with capital letter. The above error messages (e.g. lines 92~96) already disclosure the cluster(s) with problem(s), so all good.